src/internal_telemetry/allocations/allocator/tracing.rs

```diff
@@ -0,0 +1,90 @@
+use std::{any::TypeId, marker::PhantomData, ptr::addr_of};
+
+use tracing::{Dispatch, Id, Subscriber};
+use tracing_subscriber::{layer::Context, registry::LookupSpan, Layer};
+
+use super::AllocationGroupToken;
+
+pub(crate) struct WithAllocationGroup {
+    pub with_allocation_group: fn(&Dispatch, &Id, AllocationGroupToken),
+}
+
+/// [`AllocationLayer`] is a [`tracing_subscriber::Layer`] that handles entering and exiting an allocation
+/// group as the span it is attached to is itself entered and exited.
+///
+/// More information on using this layer can be found in the examples, or directly in the
+/// `tracing_subscriber` docs, found [here][tracing_subscriber::layer].
+#[cfg_attr(docsrs, doc(cfg(feature = "tracing-compat")))]
+pub struct AllocationLayer<S> {
+    ctx: WithAllocationGroup,
+    _subscriber: PhantomData<fn(S)>,
```

Comment:
I think the `fn` here is superfluous:
```suggestion
    _subscriber: PhantomData<S>,
```